(towards 598) OpenMP lowering infers firstprivate clause for some cases

sergisiso opened the pull request:
This particularly solves the issues found in NEMO on which a firstprivate clause is needed to generate functionally equivalent code to the serial version.

sergisiso commented:
@arporter This is also ready for the first review

arporter commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -537,38 +559,64 @@ def test_directive_get_private(monkeypatch):
+        subroutine my_subroutine()
+            integer :: i, scalar1, scalar2
+            real, dimension(10) :: array
+            do i = 1, 10
```
Strictly, you need to give `scalar1` a value before the loop.

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -537,38 +559,64 @@ def test_directive_get_private(monkeypatch):
+    pvars, fpvars = directive._get_private_clauses()
+    assert len(pvars.children) == 2
+    assert len(fpvars.children) == 1
+    assert pvars.children[0].name in ('i', 'scalar2')
```
I thought we sorted them so that the order was reproducible?

sergisiso replied:
True, simplified code

arporter commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -444,6 +444,99 @@ def create_loops(body):
 
+def test_omplooptrans_apply_firstprivate(fortran_reader, fortran_writer):
+    ''' Test applying the OMPLoopTrans in cases where a firstprivate
+    clause it is needed to generate functionally equivalent code than
```
"...generate code that is functionally equivalent to the original, serial version."

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -444,6 +444,99 @@ def create_loops(body):
+        subroutine my_subroutine()
+            integer :: ji, jj, jk, jpkm1, jpjm1, jpim1, scalar1, scalar2
+            real, dimension(10, 10, 10) :: zwt, zwd, zwi, zws
+            do jk = 2, jpkm1, 1
```
Initial value for `scalar1` please (in case we ever get as far as doing analysis that tells us it is never set).

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -444,6 +444,99 @@ def create_loops(body):
+!$omp end parallel do\n'''
+
+    gen = fortran_writer(psyir.children[0].children[0])
+    assert expected == gen
```
Please extend this to do compilation (even though the OMP directives will be ignored for now).

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -444,6 +444,99 @@ def create_loops(body):
+    assert expected == gen
+
+
+@pytest.mark.xfail(reason="Issue #598: Firstprivate still misses some cases")
```
Rupert and I have been finding that having an 'xfail' at this level can result in the masking of other problems. Please could you move it inside the test and trigger it if the `firstprivate` clause isn't found. i.e.:

    if expected != gen:
        pytest.xfail("...")

sergisiso replied:
I moved it inside. Note that the remaining issue is not that the firstprivate is not found, is that the DependencyAnalysis thinks it is a reduction and produces a TransformationError.

In fact I realized that with "force":True the test succeeds, so I moved the string comparison to the previous test and transform it with the force option.

arporter commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -444,6 +444,99 @@ def create_loops(body):
+do jk = 2, jpkm1, 1
+  do jj = 2, jpjm1, 1
+    do ji = 2, jpim1, 1
+      if (.true.) then
```
In fact, this test will already fail because this line doesn't correspond to the input source ;-)

sergisiso replied:
Ups, now it is properly checked in the previous test thanks to the force option.

arporter commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -444,6 +444,99 @@ def create_loops(body):
+
+    gen = fortran_writer(psyir.children[0].children[0])
+    assert expected == gen
+
```
Please also extend to do compilation.

sergisiso replied:
Done

arporter reviewed (changes requested):
Thanks for this Sergi.
It all looks good apart from the change in behaviour of the `omp parallel` directive which no longer gets a `private` clause. Please could you explain the reasoning behind this.
Apart from that, tests and examples are all fine.
Ref. guide builds without new warnings.
This is a 'towards' so I've not searched for the issue number.
I'll check pylint next time.

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -446,6 +448,7 @@ def create(children=None):
         instance.addchild(OMPDefaultClause(clause_type=OMPDefaultClause.
```
Done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -534,13 +537,19 @@ def gen_code(self, parent):
+        clauses_str = self.default_clause._clause_string
         private_list = []
-        for child in self.private_clause.children:
+        for child in private_clause.children:
```
Done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -534,13 +537,19 @@ def gen_code(self, parent):
-        private_str = "private(" + ",".join(private_list) + ")"
+        if private_list:
+            clauses_str += ", private(" + ",".join(private_list) + ")"
+        fprivate_list = []
```
Done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -583,11 +592,17 @@ def lower_to_language_level(self):
+        for child in self.children:
+            child.lower_to_language_level()
+        private_clause, fprivate_clause = self._get_private_clauses()
+        private, fprivate = self._get_private_clauses()
```
indeed, done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -665,45 +674,76 @@ def _get_private_clause(self):
+                        name = str(signature).lower()
+                        symbol = access.node.scope.symbol_table.lookup(name)
+
+                        # It is written conditionally (so we defensively use
```
Done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -665,45 +674,76 @@ def _get_private_clause(self):
-        list_result = list(result)
-        list_result.sort(key=lambda x: x[0])
+        list_private = list(private)
+        list_private.sort(key=lambda x: x[0])
```
Done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -665,45 +674,76 @@ def _get_private_clause(self):
 
-        # Create the OMPPrivateClause corresponding to the results
+        # Create the OMPPrivateClause and OMPFirstpirvateClause
         priv_clause = OMPPrivateClause()
```
Done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -1272,33 +1316,35 @@ def gen_code(self, parent):
+        # Set default() private() and firstprivate() clauses
         default_str = self.children[1]._clause_string
+        private, fprivate = self._get_private_clauses()
         private_list = []
```
Done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -1272,33 +1316,35 @@ def gen_code(self, parent):
             private_list.append(child.symbol.name)
         private_str = "private(" + ",".join(private_list) + ")"
+        fprivate_list = []
+        for child in fprivate.children:
```
Done

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -1320,18 +1366,17 @@ def lower_to_language_level(self):
         :rtype: :py:class:`psyclone.psyir.node.Node`
 
         '''
```
Yes, all but the last line. I simplified it now.

sergisiso commented on src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:
```diff
@@ -3598,6 +3598,7 @@ def test_reprod_view(monkeypatch, annexed, dist_mem):
     ompdo = colored("OMPDoDirective", Directive._colour)
     ompdefault = colored("OMPDefaultClause", Directive._colour)
     ompprivate = colored("OMPPrivateClause", Directive._colour)
```
Done

sergisiso commented on src/psyclone/tests/nemo/transformations/openmp/openmp_test.py:
```diff
@@ -115,7 +115,7 @@ def test_omp_private_declaration():
     # the node type check in order to apply the omp parallel transform.
     omp_parallel.apply(schedule.children[0:2], {'node-type-check': False})
-    expected = "!$omp parallel default(shared), private(ji,jj,jk)"
+    expected = "!$omp parallel default(shared)\n"
```
I don't mind to much which way it is but note that in this case the test docstring was checking for this in particular. So it was probably an update that brought the "private" in and we didn't read the docstring carefully.

sergisiso commented on src/psyclone/tests/nemo/transformations/openmp/openmp_test.py:
```diff
@@ -152,8 +152,7 @@ def test_omp_parallel_multi():
     # gives elements 2-3).
     otrans.apply(schedule[0].loop_body[2:4])
     gen_code = str(psy.gen).lower()
-    assert ("    !$omp parallel default(shared), private(ji,jj,zabe1,zcof1,"
```
I am happy to rethink this, but note that this is not new. In the _get_private_clause there was already this comment:
```
# Check if the write access is inside the parallel loop. If
# the write is outside of a loop, it is an assignment to
# a shared variable. Example where jpk is likely used
# outside of the parallel section later, so it must be
# declared as shared in order to have its value in other loops:
# !$omp parallel
# jpk = 100
# !omp do
# do ji = 1, jpk

# TODO #598: improve the handling of scalar variables.
```
and this is also hinted as several places, e.g: src/psyclone/tests/nemo/tests_files/explicit_do_two_loops.f90
and the #598 starting comment.

sergisiso replied:
Maybe @hiker can provide some context? Was this the behavior at some point or just marking a potential issue?

arporter replied:
I suppose this is a bit of a weird test case - we're putting a (nested) loop inside a `parallel` region but are not parallelising the loop. Previously we were at least saying that the loop variables were private (which seems sensible). I suppose the question is, why are we now saying that the loop variables are shared and should we change this?

arporter replied:
I've checked the OMP spec. and I think I'm right in saying that what we currently generate in this example will mean that the loop variables are shared between threads.

hiker replied:
I would have thought the opposite:
Section 5.1:
> Loop iteration variables inside parallel, teams, or task generating constructs are private in the innermost such construct that encloses the loop.

That would need a bit of verification/a lawyer to read the details. A similar statement is above for C/C++ (afaik Fortran kind of requires any loop variable to be private, since the number of iterations is pre-determined at the start of the loop - as in a non-parallelised loop in an inner region).

Having said that, I would actually prefer to have all loop variables listed as private, since this avoid the head scratching if this is a bug or not :)

sergisiso replied:
So you are kind of both right. I have been playing with it and the loop variable indeed defaults to private and is not a problem in the previous example. But "would be private" variables inside the loop are shared with the current implementation, so in:
```
omp parallel
scalar1 = 1
do i= start, stop
    scalar2 = <compute>
    array(i) = scalar2
enddo
omp end do
```
there is a race condition on scalar2 and the results will end up different than in the serial semantics. So I propose to keep top-level scalar reference as shared (as all threads write the same) and loop/ifs bodies scalar assignments as private (I can add a transformationError if they are used anywhere after the parallel region)

@hiker, @arporter What do you think?

arporter replied:
That sounds sensible to me. When you say 'if bodies', presumably you only mean within loops? Also, this doesn't have implications for our ability to identify reductions does it?

arporter replied:
(And thanks @hiker for managing to read the spec more carefully than I did!)

sergisiso commented on src/psyclone/tests/psyir/backend/psyir_openmp_test.py:
```diff
@@ -79,7 +79,7 @@ def test_nemo_omp_parallel(fortran_reader):
     fvisitor = FortranWriter()
     result = fvisitor(schedule)
-    correct = '''!$omp parallel default(shared), private(a,i)
+    correct = '''!$omp parallel default(shared)
```
See previous comment

sergisiso replied:
Reverted

sergisiso commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -147,11 +148,21 @@ def test_ompparallel_changes_gen_code():
+    code = str(psy.gen).lower()
+    assert "private(cell,k)" in code
 
+    # Monkeypath a case with private and firstprivate clauses
```
Fixed

sergisiso commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -182,15 +191,21 @@ def test_omp_paralleldo_changes_gen_code():
     assert "private(k)" in code
+    assert "firstprivate" not in code
+
+    # Monkeypath a case with firstprivate clauses
```
Fixed

sergisiso commented:
@arporter This is ready for review, I addressed all comments but the ones regarding the private variables only being inside the "do" construct or not.

That change would be relatively simple (just removing the worksharing_directive ancestor search in _get_private_clauses()) but the associated comment says: "Example where jpk is likely used outside of the parallel section later, so it must be declared as shared in order to have its value in other". This PR does not fix uses after the loop, so it may be safer to keep is at shared for the moment to maintain this values after the loops. But I agree it needs a better solution in the longer term, probably expanding the dependency analysis to all statements after the loop.

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -441,10 +441,10 @@ def create(children=None):
 
         instance = OMPParallelDirective(children=children)
```
pylint says that there are a couple of unused imports in this file.

sergisiso replied:
Removed

arporter commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -538,16 +538,12 @@ def gen_code(self, parent):
         zero_reduction_variables(calls, parent)
 
         clauses_str = self.default_clause._clause_string
```
Please could you disable the pylint warning here or is there a getter we can use instead?

arporter commented on src/psyclone/tests/psyir/nodes/omp_clause_test.py:
```diff
@@ -151,6 +151,37 @@ def test_private_clause():
     assert private.clause_string == "private"
 
 
+def test_private_clause_create():
```
I think you can parameterise this test so that it works for both `OMPPrivateClause` and `OMPFirstPrivateClause`. We can then do away with the almost identical test?

sergisiso replied:
Done

arporter commented on src/psyclone/tests/psyir/transformations/transformations_test.py:
```diff
@@ -444,6 +443,135 @@ def create_loops(body):
+    ''' Test applying the OMPLoopTrans in cases where a firstprivate
+    clause it is needed to generate functionally equivalent code than
+    the starting serial version.
+    
```
Trailing whitespace and unused `fortran_writer` argument.

sergisiso replied:
Removed

arporter reviewed (changes requested):
Getting there now. Aside from some pylint complaints, the main issue is to what we do with loop variables inside a parallel region. Once we've resolved that then this is good to go. I'll prompt @hiker to take a look.
Tests are still all OK.
I've now checked for pylint (see inline).

hiker commented:
I agree that for correctness we might have to parse the whole rest (or even all) of the subroutine. That might not be as bad as it sounds, since the variable accesses keep track of the statement number (called location). So we could select all access for the subroutine, identify the statement number of the first and last statements to be parallelised (though that would need some support - the accesses have the node, but iirc that's typically the Reference, NOT the statement in which it occurs, so we would need to go up the tree to find the statement)
I wonder if we could just add lastprivate for all privates, and ignore a small performance loss?

sergisiso commented:
I think it needs to be a bit more nuanced that adding lastprivate. eg. in:
```
found=False
do i=strart, stops
    if (condition) found=True
enddo
```
Adding a region with a last private it would just return the value of the last thread to execute. While in this case it is really a shared scalar.

There is also the case where we read the value of the loop variable after it, e.g:
```
do i=strart, stops
   array(i) = ...
enddo
my_val = i
```
in this case shared and lastprivate would both be wrong (the first messes the loop and the second may change the value of i depending on the order of execution). In this case we need a new assingment:
```
omp parallel do
  do i=strart, stops
     array(i) = ...
  enddo
end ompparallel do
i = stops
my_val = i
```

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -441,10 +441,10 @@ def create(children=None):
-        # and an OMPPrivateClause.
+        # An OMPParallelDirective must have 4 children.
+        # Child 0 is a Schedule, created in the constructor.
+        # The create function adds the other three mandatory children: 
```
Removed

sergisiso commented on src/psyclone/psyir/nodes/omp_directives.py:
```diff
@@ -1320,15 +1310,12 @@ def gen_code(self, parent):
         # Set default() private() and firstprivate() clauses
         default_str = self.children[1]._clause_string
```
I went with disable warning because hopefully we will get rid of gen_code soon

arporter replied:
This one is still there?

sergisiso replied:
Ups, last time I disabled the warning for another _clause_string access in the same file. Now are both warnings disabled!

arporter replied:
Thanks :-)

sergisiso commented:
@arporter This is ready for another review. Now scalar writes inside (any) loop imply the variable is private.

arporter commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -166,12 +177,10 @@ def test_omp_paralleldo_changes_gen_code():
     assert isinstance(tree.children[0], OMPParallelDoDirective)
     pdir = tree.children[0]
     code = str(psy.gen).lower()
-    assert len(pdir.children) == 4
```
I'm sure I should know the answer to this but why does the OMPParallelDoDirective now only have two children where before it had four?

sergisiso replied:
The answer may be subtle but it is because before there was some attempt at caching/storing the computed clauses once they were computed (e.g. in the gen_code). I am trying to get rid of these things as much as possible because we don't invalidate the computed values!

So if we modified the tree after a first gen_code and added another private symbol, the next gen_code would not recompute the clauses and before this was generating incorrect code.

For the same reason I am now always recomputing the _get_private_clauses in the lower_to_language_level even if they already exist.

arporter replied:
Aha. Thanks for explaining.

arporter reviewed (changes requested):
Very nearly there now.
Just a couple of pylint things and one question about the number of children.
All tests and examples are fine.
Ref. guide still builds fine.

sergisiso commented on src/psyclone/tests/psyir/nodes/omp_directives_test.py:
```diff
@@ -1,7 +1,7 @@
 # -----------------------------------------------------------------------------
```
Removed them

sergisiso commented:
@arporter Ready for another look
